baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/StreamFlatMapOptional.java

```diff
@@ -30,50 +32,132 @@
+                    .named("map")
+                    .withParameters(Function.class.getName()),
             // Any of the three MatchTypes are reasonable in this case, given a single arg
-            MatchType.LAST,
-            OPTIONAL_STREAM);
+            MatchType.AT_LEAST_ONE,
+            MethodMatchers.instanceMethod()
+                    .onDescendantOf(Optional.class.getCanonicalName())
+                    .namedAnyOf("get", "orElseThrow")
+                    .withNoParameters());
 
     @Override
     public Description matchMethodInvocation(MethodInvocationTree tree, VisitorState state) {
         if (STREAM_FLATMAP_OPTIONAL_STREAM.matches(tree, state)) {
-            ExpressionTree receiver = ASTHelpers.getReceiver(tree.getMethodSelect());
-            if (receiver != null) {
-                SuggestedFix.Builder fix = SuggestedFix.builder();
-                String optionalType = SuggestedFixes.qualifyType(state, fix, Optional.class.getCanonicalName());
-                String replacement = ".filter(" + optionalType + "::isPresent).map(" + optionalType + "::get)";
-                return buildDescription(tree)
-                        .addFix(fix.replace(state.getEndPosition(receiver), state.getEndPosition(tree), replacement)
-                                .build())
-                        .build();
+            ExpressionTree methodSelect = tree.getMethodSelect();
+            ExpressionTree receiver = ASTHelpers.getReceiver(methodSelect);
+            return fix(tree, state, receiver, receiver);
+        }
+
+        if (STREAM_MAP_GET.matches(tree, state)) {
+            ExpressionTree mapTree = ASTHelpers.getReceiver(tree.getMethodSelect());
+            if (mapTree != null && STREAM_FILTER_IS_PRESENT.matches(mapTree, state)) {
+                ExpressionTree filterTree = ASTHelpers.getReceiver(mapTree);
+                return fix(tree, state, tree.getMethodSelect(), filterTree);
             }
         }
+
         return Description.NO_MATCH;
     }
+
+    private Description fix(
+            MethodInvocationTree tree, VisitorState state, ExpressionTree receiver, ExpressionTree expressionTree) {
+        if (receiver == null) {
+            return Description.NO_MATCH;
+        }
+
+        Type elementType = ASTHelpers.getType(tree);
+        if (elementType == null) {
+            return Description.NO_MATCH;
+        }
+
+        Type receiverType = ASTHelpers.getType(receiver);
+        if (receiverType == null) {
+            return Description.NO_MATCH;
+        }
+
+        SuggestedFix.Builder fix = SuggestedFix.builder();
+
+        long receiverCount = args(ASTHelpers.getReceiverType(receiver)).count();
+        Stream<Type> args = args(ASTHelpers.getReceiverType(expressionTree));
+        long argCount = args.count();
+        boolean shouldQualifyType =
+                args(elementType).findAny().isPresent() || receiverCount > 1 || (receiverCount == 0 && argCount > 0);
```

Comment:
Worth running this on a couple larger internal projects to verify the results compile successfully. If there are edge cases we can't easily isolate, I think it should be safe to over-qualify.

Reply:
Yeah, this logic isn't quite right yet. I've tested out on a few local repos and there are some places were I'm not generating qualified types that need them.

I'm trying to avoid adding unnecessary qualifying type args, but there are at least a few cases I haven't quite nailed down that need qualification (e.g. some stream chains extracting layers of optionals).

Reply:
> I'm trying to avoid adding unnecessary qualifying type args, but there are at least a few cases I haven't quite nailed down that need qualification (e.g. some stream chains extracting layers of optionals).

Thinking about this a bit more, I think it's actually a good fit for the incredibly expensive `SuggestedFixes.compilesWithFix` utility, because in the failing case, we can qualify the method reference, and needn't pay the cost again (unlike `LambdaMethodReference` which used `compilesWithFix` in the matching component, meaning it must run the test in every subsequent compilation).